physical/postgresql/postgresql.go

```diff
@@ -213,3 +282,198 @@ func (m *PostgreSQLBackend) List(ctx context.Context, prefix string) ([]string,
 
 	return keys, nil
 }
+
+// LockWith is used for mutual exclusion based on the given key.
+func (p *PostgreSQLBackend) LockWith(key, value string) (physical.Lock, error) {
+	identity, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, err
+	}
+	return &PostgreSQLLock{
+		backend:  p,
+		key:      key,
+		value:    value,
+		identity: identity,
+	}, nil
+}
+
+func (p *PostgreSQLBackend) HAEnabled() bool {
+	return p.haEnabled
+}
+
+// Lock tries to acquire the lock by repeatedly trying to create
+// a record in the PostgreSQL table. It will block until either the
+// stop channel is closed or the lock could be acquired successfully.
+// The returned channel will be closed once the lock is deleted or
+// changed in the PostgreSQL table.
+func (l *PostgreSQLLock) Lock(stopCh <-chan struct{}) (doneCh <-chan struct{}, retErr error) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	if l.held {
+		return nil, fmt.Errorf("lock already held")
+	}
+
+	done := make(chan struct{})
+	// close done channel even in case of error
+	defer func() {
+		if retErr != nil {
+			close(done)
+		}
+	}()
+
+	var (
+		stop    = make(chan struct{})
+		success = make(chan struct{})
+		errors  = make(chan error)
+		leader  = make(chan struct{})
+	)
+	// try to acquire the lock asynchronously
+	go l.tryToLock(stop, success, errors)
+
+	select {
+	case <-success:
+		l.held = true
+		// after acquiring it successfully, we must renew the lock periodically,
+		// and watch the lock in order to close the leader channel
+		// once it is lost.
+		go l.periodicallyRenewLock(leader)
```

Comment:
It feels like `watch` and `periodicallyRenewLock` are needlessly complicated.  `watch` periodically looks to see if we lost the lock, and if we did, stops `periodicallyRenewLock`.  `periodicallyRenewLock` periodically updates the lock, ignoring errors.  Can we simplify this by eliminating `watch` and making `periodicallyRenewLock` exit on failure to renew the lock?

Reply:
I've not reviewed this PR, but the functions you mention were related to DynamoDB code that was just updated to fix a race condition, so fair warning.  https://github.com/hashicorp/vault/issues/5828

Reply:
> I've not reviewed this PR, but the functions you mention are related to DynamoDB code that was just updated to fix a race condition, so fair warning. #5828

Thanks for the reference @kalafut .  I believe the race condition applies to this PR's code as well.  The crux of the issue seems to be that Unlock() has a contract to give up the lock, but we keep trying to re-acquire it.  I think Unlock should cause renewal attempts to cease.  One way to do this would be to put the ticker used by the renewal func into the lock object, and have Unlock call Stop() on it.  

I still feel this is easier to understand if watch() is out of the picture.

Reply:
As stated above, My aim was to keep dynamoDB logic, but I too found these somewhat overlapping. I my mind there is no need to have two tickers trying to steal or guard lock, one should be sufficient as you suggest.